### ISSUE TITLE
Cannot connect to a Remote Tunnel for a second time from local vscode 

### ISSUE THREAD

**Issue by OscarLlorente** [microsoft/vscode#168735]
Does this issue occur when all extensions are disabled?: The remote tunnel extension is needed

- VS Code Version: 1.74.0
- OS Version: Windows 10

Steps to Reproduce:

1. Set up tunneling on machine 1
2. Connect from local vscode in machine 2 to machine 1
3. Close remote from machine 2
4. Connect again from local vscode in machine 2 to machine 1

The following can be seen in machine 1 terminal:
[2022-12-10 17:35:05] info [rpc.6] Found running server (pid=3124)
[2022-12-10 17:35:15] info [rpc.6] Forwarding port 5555
[2022-12-10 17:35:41] info [rpc.6] Disposed of connection to running server.
[2022-12-10 17:38:37] info [tunnels::connections::relay_tunnel_host] Opened new client on channel 2

And it never establishes the connection

**Comment by VSCodeTriageBot** [comment]
Thanks for creating this issue! It looks like you may be using an old version of VS Code, the latest stable release is 1.74.0. Please try upgrading to the latest version and checking whether this issue remains.

Happy Coding!

**Comment by OscarLlorente** [comment]
Sorry just typo error, I am using 1.74.0 the lastest version available

**Comment by OscarLlorente** [comment]
Kubeflow is the name that I gave the remote machine to create the tunnel:

```
2022-12-12 17:18:10.295 [info] Resolving tunnel+kubeflow...
2022-12-12 17:18:10.336 [info] Found running server in \\.\pipe\vscode-tunneling-kubeflow-44eb4f1ef5299310c91fed60dbd51aaa29fb6f2d-sock
2022-12-12 17:18:10.341 [info] Remote resolution completed
2022-12-12 17:18:10.350 [debug] [proxy] [connection.4] connection from:  127.0.0.1 59603
2022-12-12 17:18:10.463 [info] [proxy] [connection.4] tunnel connection established
2022-12-12 17:18:10.531 [debug] [proxy] [connection.5] connection from:  127.0.0.1 59603
2022-12-12 17:18:10.643 [info] [proxy] [connection.5] tunnel connection established
2022-12-12 17:19:12.390 [info] Resolving tunnel+kubeflow...
2022-12-12 17:19:12.448 [info] Creating new tunnel proxy server
2022-12-12 17:19:16.227 [info] [proxy] Found tunnel (ID=h4lw9hms, Cluster=euw)
2022-12-12 17:19:16.234 [info] Remote resolution completed
2022-12-12 17:19:16.267 [debug] [proxy] [connection.0] connection from:  127.0.0.1 59795
2022-12-12 17:19:17.474 [debug] [proxy] Tunnel started connecting
2022-12-12 17:19:17.514 [info] Resolving tunnel+kubeflow...
2022-12-12 17:19:17.515 [info] Creating new tunnel proxy server
2022-12-12 17:19:21.223 [info] [proxy] Found tunnel (ID=h4lw9hms, Cluster=euw)
2022-12-12 17:19:21.227 [info] Remote resolution completed
2022-12-12 17:19:21.234 [debug] [proxy] [connection.0] connection from:  127.0.0.1 59806
2022-12-12 17:19:22.309 [debug] [proxy] Tunnel started connecting
2022-12-12 17:19:22.329 [info] Resolving tunnel+kubeflow...
2022-12-12 17:19:22.330 [info] Creating new tunnel proxy server
2022-12-12 17:19:25.908 [info] [proxy] Found tunnel (ID=h4lw9hms, Cluster=euw)
2022-12-12 17:19:25.914 [info] Remote resolution completed
2022-12-12 17:19:25.931 [debug] [proxy] [connection.0] connection from:  127.0.0.1 59815
2022-12-12 17:19:27.050 [debug] [proxy] Tunnel started connecting
2022-12-12 17:19:27.064 [info] Resolving tunnel+kubeflow...
2022-12-12 17:19:27.064 [info] Creating new tunnel proxy server
2022-12-12 17:19:30.621 [info] [proxy] Found tunnel (ID=h4lw9hms, Cluster=euw)
2022-12-12 17:19:30.627 [info] Remote resolution completed
2022-12-12 17:19:30.638 [debug] [proxy] [connection.0] connection from:  127.0.0.1 59822
2022-12-12 17:19:31.871 [debug] [proxy] Tunnel started connecting
2022-12-12 17:19:31.888 [info] Resolving tunnel+kubeflow...
2022-12-12 17:19:31.889 [info] Creating new tunnel proxy server
2022-12-12 17:19:35.379 [info] [proxy] Found tunnel (ID=h4lw9hms, Cluster=euw)
2022-12-12 17:19:35.381 [info] Remote resolution completed
2022-12-12 17:19:35.386 [debug] [proxy] [connection.0] connection from:  127.0.0.1 59837
2022-12-12 17:19:36.484 [debug] [proxy] Tunnel started connecting
```

**Comment by disberd** [comment]
I also have basically the same issue on 1.74.3.

For me the trace log hangs at the line like:
`2023-01-15 12:08:27.047 [debug] [proxy] [connection.3] connection from:  127.0.0.1 63904`
every time I try to reload the window.

The connection works fine from the web interface (so not from local vscode but directly from the browser) and I am able to re-establish the connection from local vscode if I restart the tunnel on the remote machine.

**Comment by ramirocarra** [comment]
Same here, It got stuck reconnecting.

The Remote keeps opening connections but it never reconnects. Port forwarding still worked fine.

I had to restart the local and remote to be able to reconnect.

Using 1.74.3 on both machines.

**Comment by ajinkya-k** [comment]
I have the exact same issue. I can connect to the host server form the web interface on my client but the local vscode on the client machine cannot connect. When i try to connect to the server machine here is what the log on the server side shows: 
```
[2023-01-20 01:18:15] info [rpc.0] Disposed of connection to running server.
[2023-01-20 09:16:03] info [tunnels::connections::relay_tunnel_host] Opened new client on channel 3
[2023-01-20 09:16:07] info [tunnels::connections::relay_tunnel_host] Error connecting forwarding to 127.0.0.1:12146, No connection could be made because the target machine actively refused it. (os error 10061)
[2023-01-20 09:16:09] info [tunnels::connections::relay_tunnel_host] Opened new client on channel 4
[2023-01-20 09:16:13] info [tunnels::connections::relay_tunnel_host] Error connecting forwarding to 127.0.0.1:12146, No connection could be made because the target machine actively refused it. (os error 10061)
[2023-01-20 09:16:14] info [tunnels::connections::relay_tunnel_host] Opened new client on channel 5
[2023-01-20 09:16:18] info [tunnels::connections::relay_tunnel_host] Error connecting forwarding to 127.0.0.1:12146, No connection could be made because the target machine actively refused it. (os error 10061)
```

**Comment by ajinkya-k** [comment]
Update: I installed the Insider version of VS code and ran a tunnel server using the command `code-insider tunnel`, while still using the stable version of VS code on the client machine and this issue mostly goes away. Any idea where I can find the changelog for this specific fix?

**Comment by ajinkya-k** [comment]
Its weird, this issue always happens when using VSCode 1.75.0 production version, but it happens only sometimes if I use VSCode Insiders 1.75.0

**Comment by ajinkya-k** [comment]
This issue still persists. I can connect form the browser on the guest machine multiple times, but not from the VScode on the guest machine.

**Comment by excelsior091224** [comment]
> This issue still persists. I can connect form the browser on the guest machine multiple times, but not from the VScode on the guest machine. I get the following error:
> 
> Failed to connect to the remote extension host server (Error: WebSocket close with status code 1006)

It seems same problem of mine.
https://github.com/microsoft/vscode/issues/178316

**Comment by blood400cc** [comment]
exact same issue here

**Comment by Interpause** [comment]
Still getting `Error: WebSocket close with status code 1006` on second connection for both the pre-release version of the Remote Tunnels extension (v1.2.2023042609) and release version (v1.1.0) as of today. <https://vscode.dev/tunnel> still works, where I can issue `code tunnel restart`. However, it doesn't seem to be effective in allowing the local VSCode install to reconnect even after restarting VSCode.

**Comment by excelsior091224** [comment]
It does not seem to have been fixed.
I am still getting "WebSocket close with status code 1006".
<img width="1050" alt="image" src="https://github.com/microsoft/vscode/assets/44745901/3c0b6677-4218-49a8-9407-82ee7ceeaac5">

**Comment by ringosham** [comment]
Same here. I've tired disabling all firewall rules for VSCode. And even reinstalling, purging all configs. Version 1.78.1

Attached log. This differs from machine to machine. It worked on another computer but not this one.
It seems to be something from Windows blocking the connection?
[vscode-app-1683730524875.log](https://github.com/microsoft/vscode/files/11443297/vscode-app-1683730524875.log)

**Comment by Zebnastien** [comment]
Same here : vstunnel connects once, then it won't reconnect until I restart code tunnel.

**Comment by PastelStoic** [comment]
I've had this issue happen twice now, on two totally separate computers (one Windows 10, one Ubuntu). Tunnel works fine for a while, then it just *stops* working, and even reinstalling vscode doesn't fix the issue. "Error: Timeout connecting to relay" is what I get, eventually. I've run out of computers, and now I can't do any work in a browser.

**Comment by alexmill** [comment]
Same issue here.

**Comment by connor4312** [comment]
"Timeout connecting to relay" is a separate issue and is being tracked by the team responsible for Dev Tunnels.

**Comment by alexmill** [comment]
I've had some good luck and some bad luck using the Remote Tunnels across recent updates of the Insiders edition. I think, however, I have found a consistent way to cause this error for me:

After opening a connection from A to B,  I start a Jupyter server running on 127.0.0.1 of machine B. VS Code conveniently forwards Jupyter's :8888 port over the tunnel and allows me to connect the kernel to the files I'm working on. This works for a couple hours, but consistently this configuration will fail after a few hours.

I was just doing this today and briefly lost connection... After trying to reconnect to the remote machine, my local machine is just hanging on "Connecting to... ". Restarting the app on the remote machine or turning remote tunnels off/on always fixes the issue. 

Previously, I have noticed the tunnel remaining active and connectable for several days at a time. But as I started working on a notebook today over the tunnel, that is when the tunnel lost connection.

**Comment by JK88-1337** [comment]
Same issue, happened after the last update.
`Ubuntu 22.04.2 LTS`
Error message:
`[2023-07-12 06:16:00] info [tunnels::connections::relay_tunnel_host] Error connecting forwarding to 127.0.0.1:26755, Connection refused (os error 111)`

**Comment by ramirocarra** [comment]
> I've had some good luck and some bad luck using the Remote Tunnels across recent updates of the Insiders edition. I think, however, I have found a consistent way to cause this error for me:
> 
> After opening a connection from A to B, I start a Jupyter server running on 127.0.0.1 of machine B. VS Code conveniently forwards Jupyter's :8888 port over the tunnel and allows me to connect the kernel to the files I'm working on. This works for a couple hours, but consistently this configuration will fail after a few hours.
> 
> I was just doing this today and briefly lost connection... After trying to reconnect to the remote machine, my local machine is just hanging on "Connecting to... ". Restarting the app on the remote machine or turning remote tunnels off/on always fixes the issue.
> 
> Previously, I have noticed the tunnel remaining active and connectable for several days at a time. But as I started working on a notebook today over the tunnel, that is when the tunnel lost connection.

I it helps. My situation is very similar, it happens when I start a TensorBoard server in port 6006 and the port is forwarded by the tunnel.

**Comment by chux0519** [comment]
Same here. Restarting can be a workaround, but it's a bit annoying.

**Comment by Superviro** [comment]
I'm suffering from this issue too, I cannot even reconnect from the web interface

**Comment by saisonxiang** [comment]
I have the same issue. Remote Tunnels says both terminals are running but it just perpetually hangs, trying to connect. I can "remote connect" one computer to itself?!!?? Are the engineers at Microsoft jerking off to the message boards!?!?!

**Comment by saisonxiang** [comment]
Same here. MacOS client to Linux host. I've wasted months trying to connect a computer 6 inches away from another. Both of which I paid for.

**Comment by connor4312** [comment]
Thanks for the hint @alexmill. I found an error state that could be hit when using port forwarding. Remote Tunnels extension version 1.4.1, releasing presently, will have the fix. Please let me know if people hit it again.

**Comment by connor4312** [comment]
Actually, I verified that it looks like the same issue that OP hit based on their logs, so I will close this issue once the release is out.

**Comment by alexmill** [comment]
Seems to be working when I tried recently. @connor4312, thank you for looking into this! You've resolved a huge headache for me

**Comment by fonewiz** [comment]
I having the same issue when connecting from Desktop VS Code client (Version: 1.81.1) running on Mac OS to a VS Code server running on Ubuntu.
![Screenshot 2023-08-30 at 2 42 14 PM](https://github.com/microsoft/vscode/assets/7445599/a3825758-28c6-4d20-8533-8e8ca07ab781)

No such error or issue if I connect to the same VS Code server from a web browser.

I haven't made any changes to the server or client side since this was last working (last night it worked), today it's busted.

Any help would be appreciated. Thanks

**Comment by roblourens** [comment]
@OscarLlorente, this issue should be fixed, please try with the latest pre-release Remote-Tunnels extension and let us know if you still see it, thanks

**Comment by connor4312** [comment]
Marking verified based on alexmill's testimony

**Comment by alexmill** [comment]
This was working for me for days at a time without issue. A couple times now, I have got into an error state where I can no longer remote into the serving machine. Maybe related to the fact that I often close my client machine (laptop) while the tunnel is active. When I open the machine later, the connection gets stuck on a loop of "connecting". 

However, also a couple times now, I have started to file a new comment here noting the issue. But both times, in the time it takes to write my comment, I have gone back to the tunnel to find that it finally succeeded in connecting. It takes maybe 10-15 minutes of retrying the connection, but it has been coming through eventually. Will follow-up if the issue persists long term.
